enarx-keep-sev/src/vm.rs

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::x86_64::*;
+
+use std::error::Error;
+
+use kvm_bindings::kvm_userspace_memory_region as MemoryRegion;
+use kvm_ioctls::{Kvm, VmFd};
+use memmap::*;
+use x86_64::structures::paging::page_table::PageTableFlags;
+use x86_64::PhysAddr;
+
+const DEFAULT_MEM_SIZE: usize = units::bytes!(1; GiB);
+
+pub struct VirtualMachine {
+    kvm: Kvm,
+    fd: VmFd,
+    address_space: Vec<MemoryRegion>,
+    mmap_handles: Vec<MmapMut>,
+}
+
+impl VirtualMachine {
+    pub fn new() -> Result<Self, Box<dyn Error>> {
+        // Create a KVM context
+        let kvm = Kvm::new()?;
+        let fd = kvm.create_vm()?;
+
+        // Create the guest address space
+        let guest_addr_space = memmap::MmapOptions::new()
+            .len(DEFAULT_MEM_SIZE)
+            .map_anon()?;
+        let region = MemoryRegion {
+            slot: 0,
+            flags: 0,
+            guest_phys_addr: 0,
+            memory_size: DEFAULT_MEM_SIZE as _,
+            userspace_addr: &*guest_addr_space as *const _ as *const u8 as u64,
+        };
+
+        unsafe {
+            fd.set_user_memory_region(region)?;
+        }
+
+        // Set up the page tables
+        let mut page_tables = PageTables::default();
+        let pdpte_start =
+            PhysAddr::try_new(PDPTE_START).map_err(|_| PhysAddrNotValid(PDPTE_START))?;
+        page_tables.pml4t[0].set_addr(
+            pdpte_start,
+            PageTableFlags::WRITABLE | PageTableFlags::PRESENT,
+        );
+
+        page_tables.pml3t_ident[0].set_flags(
+            PageTableFlags::HUGE_PAGE | PageTableFlags::WRITABLE | PageTableFlags::PRESENT,
+        );
```

Comment:
I think this helps us keep the page table structures smaller, since bigger pages would result in fewer entries.

Reply:
Makes sense. I am not sure how this translates to the NPT mappings - does this map to an actual huge page on the host ? I guess at this stage, we  are not worried about that but this can also lead to a lot of wasted host memory (unless we are actually using that memory, ofcourse!)

Reply:
I don't know what NPT is, but all page table mappings here should be confined to the big block of host memory I mmap'd to serve as the guest's address space. I don't think the host will be impacted beyond having to supply that big block of memory for the address space.

Reply:
They should be allocated via `mmap()` with `MAP_HUGE_2MB` on the host side. Again, it would be good if we can enforce type safety for this.